src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,682 @@
+            #[inline]
+            pub fn checked_assign(&mut self, x: $x) -> Result<(), $crate::CheckedAssignError> {
+                if x <= Self::MAX {
+                    let shift = x << START;
+                    *self.0 = shift | (*self.0 & !Self::MASK);
+                    Ok(())
+                } else {
+                    Err($crate::CheckedAssignError)
+                }
+            }
+            #[doc = concat!("
+                Conveniance alias for [`Self::MAX`].
+                
+                ```
+                use bit_fields::BitRangeMut;
+                let mut x = ", stringify!($x), "::default();
+
+                assert_eq!(BitRangeMut::<_,0,1>(&mut x).get_max(),1);
+                assert_eq!(BitRangeMut::<_,0,2>(&mut x).get_max(),3);
+                assert_eq!(BitRangeMut::<_,0,3>(&mut x).get_max(),7);
+                assert_eq!(BitRangeMut::<_,0,4>(&mut x).get_max(),15);
+                assert_eq!(BitRangeMut::<_,0,5>(&mut x).get_max(),31);
+                assert_eq!(BitRangeMut::<_,0,6>(&mut x).get_max(),63);
+                assert_eq!(BitRangeMut::<_,0,7>(&mut x).get_max(),127);
+                assert_eq!(BitRangeMut::<_,0,8>(&mut x).get_max(),255);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_max(&self) -> $x {
+                Self::MAX
+            }
+            #[doc = concat!("
+                Conveniance alias for [`Self::MASK`].
+                
+                ```
+                use bit_fields::BitRangeMut;
+                let mut x = ", stringify!($x), "::default();
+
+                assert_eq!(BitRangeMut::<_,0,1>(&mut x).get_mask(),1);
+                assert_eq!(BitRangeMut::<_,1,3>(&mut x).get_mask(),6);
+                assert_eq!(BitRangeMut::<_,3,7>(&mut x).get_mask(),120);
+                assert_eq!(BitRangeMut::<_,7,8>(&mut x).get_mask(),128);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_mask(&self) -> $x {
+                Self::MASK
+            }
+        }
+    };
+}
+
+bit_mut_range!(u128, mask_u128, max_u128);
+bit_mut_range!(u64, mask_u64, max_u64);
+bit_mut_range!(u32, mask_u32, max_u32);
+bit_mut_range!(u16, mask_u16, max_u16);
+bit_mut_range!(u8, mask_u8, max_u8);
+
+// impl From<..> for BitRange<..>
+macro_rules! bit_range_from {
+    ($x:ty) => {
+        // These values are only evaluated at compile-time, thus a failure can only occur at
+        // compile-time and would be immediately obvious. Thus it is safe to use arithmetic here.
+        #[allow(clippy::integer_arithmetic, clippy::arithmetic)]
+        impl<const START: u8, const END: u8> From<&BitRange<'_, $x, START, END>> for $x {
+            #[inline]
+            fn from(this: &BitRange<'_, $x, START, END>) -> Self {
+                let a = BitRange::<'_, $x, START, END>::MASK & *this.0;
+                a >> START
+            }
+        }
```

Comment:
I don't think  the comment applies here, as `from` is not `const fn`, but I also don't think a right shift would cause any troubles anyway

Reply:
I believe right shift can panic if trying to shift by more bits than are present in the type (e.g. `START > 8 * size_of_val(a)`) but in this case since the size of the type of the amount of shift are known at compile time a failure can only occur at compile time.

Reply:
I see, slightly odd behavior on Rust's part but oh well. However, I am still a bit confused what you mean by "evaluated at compile-time", as the body of `from` is not const (Although even if it were, if you call a const function from a non-const context, it'll get evaluated at runtime). The following test compiles passes (as in, it hits a panic)

```rs
#[should_panic]
#[test]
fn invalid_bit_access() {
    let field = BitFieldu128(0);
    let range = BitRange::<u128, 150, 130>(&field.0);

    let result = u128::from(&range);
}
```

Reply:
Never mind I was wrong. I will adjust the comment accordingly.